recipe/meta.yaml

```diff
@@ -136,32 +110,30 @@ requirements:
     # Otherwise build and host might end up with two different versions
     # If they do, merge_build_host will fail.
     - readline                           # [linux]
+    # For QDoc
+    # http://doc.qt.io/Qt-5/qdoc-guide-clang.html
+    # Only libclang is needed, but needs the full package for detection.
+    - llvm
+    - llvmdev
+    - clangdev
+    - libclang
```

Comment:
Why were they moved here? They are host dependencies not build dependencies.

Reply:
@andfoy, why was this conversation marked as resolved? I already marked it unresolved multiple times.

Reply:
It seems that setting `LLVM_INSTALL_DIR` to point to `PREFIX` causes several compilation errors on arm64. So it works fine to have those dependencies on the build section

Reply:
What errors?

Reply:
```
ld: warning: ignoring file /Users/macmini/Documents/qt-feedstock/miniforge3/conda-bld/qt_1636756961995/_h_env_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_pl
/lib/libclang.dylib, building for macOS-x86_64 but attempting to link with file built for macOS-arm64
ld: warning: ignoring file /Users/macmini/Documents/qt-feedstock/miniforge3/conda-bld/qt_1636756961995/_h_env_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_pl
/lib/libc++.dylib, building for macOS-x86_64 but attempting to link with file built for macOS-arm64
Undefined symbols for architecture x86_64:                                                                                                      
  "___gxx_personality_v0", referenced from:                             
      -[QMacAutoReleasePoolTracker dealloc] in libQt5Bootstrap.a(qcore_mac.o)                                                                
      QMacAutoReleasePool::~QMacAutoReleasePool() in libQt5Bootstrap.a(qcore_mac.o)                                                             
      QMacVersion::applicationVersion() in libQt5Bootstrap.a(qcore_mac.o)                                                                       
      QMacVersion::libraryVersion() in libQt5Bootstrap.a(qcore_mac.o)                                                                        
      Dwarf Exception Unwind Info (__eh_frame) in libQt5Bootstrap.a(qcore_mac.o)                                                             
      QString::fromCFString(__CFString const*) in libQt5Bootstrap.a(qcore_foundation.o)                                                         
      QString::fromNSString(NSString const*) in libQt5Bootstrap.a(qcore_foundation.o)                                                           
      ...                                                               
  "vtable for __cxxabiv1::__vmi_class_type_info", referenced from:      
      typeinfo for WebXMLGenerator in webxmlgenerator.o                                                                                         
  NOTE: a missing vtable usually means the first non-inline virtual member function has no definition.                                          
  "operator new[](unsigned long)", referenced from:                     
      editDistance(QString const&, QString const&) in editdistance.o                                                                         
      Tokenizer::init() in tokenizer.o                                                                                                          
      QHashData::detach_helper(void (*)(QHashData::Node*, void*), void (*)(QHashData::Node*), int, int) in libQt5Bootstrap.a(qhash.o)                                                                                                                                                     
      QHashData::rehash(int) in libQt5Bootstrap.a(qhash.o)                                                                                                                                                                                                                                
      QDirPrivate::sortFileList(QFlags<QDir::SortFlag>, QList<QFileInfo>&, QStringList*, QList<QFileInfo>*) in libQt5Bootstrap.a(qdir.o)                                                                                                                                                  
      QMessagePattern::setPattern(QString const&) in libQt5Bootstrap.a(qlogging.o)                                                           
  "operator delete[](void*)", referenced from:                          
      editDistance(QString const&, QString const&) in editdistance.o                                                                         
      Tokenizer::~Tokenizer() in tokenizer.o                          
      QHashData::free_helper(void (*)(QHashData::Node*)) in libQt5Bootstrap.a(qhash.o)                                                       
      QHashData::rehash(int) in libQt5Bootstrap.a(qhash.o)            
      QScopedPointerArrayDeleter<QDirSortItem>::cleanup(QDirSortItem*) in libQt5Bootstrap.a(qdir.o)                                          
      std::__1::enable_if<_CheckArrayPointerConversion<char const**>::value, void>::type std::__1::unique_ptr<char const* [], std::__1::default_delete<char const* []> >::reset<char const**>(char const**) in libQt5Bootstrap.a(qlogging.o)
      std::__1::default_delete<std::__1::unique_ptr<char const [], std::__1::default_delete<char const []> > []>::_EnableIfConvertible<std::__1::unique_ptr<char const [], std::__1::default_delete<char const []> > >::type std::__1::default_delete<std::__1::unique_ptr<char const [], 
std::__1::default_delete<char const []> > []>::operator()<std::__1::unique_ptr<char const [], std::__1::default_delete<char const []> > >(std::__1::unique_ptr<char const [], std::__1::default_delete<char const []> >*) const in libQt5Bootstrap.a(qlogging.o)
      ...                                                             
  "___cxa_guard_acquire", referenced from:                            
      Singleton<Config>::instance() in codeparser.o                   
      CppCodeMarker::addMarkUp(QString const&, Node const*, Location const&) in cppcodemarker.o                                              
      QGlobalStatic<QHash<QString, int>, (anonymous namespace)::Q_QGS_cmdHash::innerFunction(), (anonymous namespace)::Q_QGS_cmdHash::guard>::operator()() in doc.o
      QGlobalStatic<QHash<QString, Macro>, (anonymous namespace)::Q_QGS_macroHash::innerFunction(), (anonymous namespace)::Q_QGS_macroHash::guard>::operator()() in doc.o
      QGlobalStatic<QMap<QString, QString>, (anonymous namespace)::Q_QGS_aliasMap::innerFunction(), (anonymous namespace)::Q_QGS_aliasMap::guard>::operator()() in doc.o
      Doc::location() const in doc.o                                  
      Doc::body() const in doc.o
```

Reply:
The stacktrace is longer

Reply:
> building for macOS-x86_64 but attempting to link with file built for macOS-arm64

This means qt is doing something wrong. It's trying to build for x86_64 and linking with an arm64 library. With your change here, it's building for x86_64 and linking with an x86_64 library. This is not what we want.

`qdoc` is built for x86_64. See the following excerpt from the logs>

> $BUILD_PREFIX/bin/x86_64-apple-darwin13.4.0-clang++ -stdlib=libc++ -headerpad_max_install_names  -arch x86_64 -isysroot /Applications/Xcode_12.4.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX11.0.sdk -mmacosx-version-min=10.13 -Wl,-dead_strip -Wl,-rpath,$BUILD_PREFIX/lib -Wl,-rpath,@executable_path/../Frameworks -o ../../bin/qdoc .obj/atom.o .obj/clangcodeparser.o .obj/codechunk.o .obj/codemarker.o .obj/codeparser.o .obj/config.o .obj/cppcodemarker.o .obj/cppcodeparser.o .obj/doc.o .obj/docbookgenerator.o .obj/editdistance.o .obj/generator.o .obj/helpprojectwriter.o .obj/htmlgenerator.o .obj/location.o .obj/main.o .obj/node.o .obj/openedlist.o .obj/parameters.o .obj/puredocparser.o .obj/qdocdatabase.o .obj/qdoctagfiles.o .obj/qdocindexfiles.o .obj/quoter.o .obj/sections.o .obj/separator.o .obj/text.o .obj/tokenizer.o .obj/tree.o .obj/xmlgenerator.o .obj/yyindent.o .obj/webxmlgenerator.o .obj/qdoccommandlineparser.o .obj/utilities.o .obj/jscodemarker.o .obj/qmlcodemarker.o .obj/qmlcodeparser.o .obj/qmlmarkupvisitor.o .obj/qmlvisitor.o   -L$BUILD_PREFIX/lib -lclang $SRC_DIR/qt-build/qtdeclarative/lib/libQt5QmlDevTools.a $SRC_DIR/qt-build/qtbase/lib/libQt5Bootstrap.a -framework Foundation -framework CoreServices

Reply:
That's because the host is a `x86_64` machine, therefore some binaries such as qmake and qdoc get built for such architecture. When `LLVM_INSTALL_DIR` points to `BUILD_PREFIX`, then no errors occur

Reply:
Right, then we can't use the binaries like `qmake`, `qdoc`, etc on the target machine unless rosetta is installed.

Reply:
Since there's no x86_64 libclang-11.so, `qdoc` will not be usable on the target machine at all.

Reply:
Can we leave it as it is? Or what do you suggest @isuruf?

Reply:
Two options,
1. Disable cross compiling arm64 and compile natively for osx-arm64
2. Build host tools with `-arch x86_64 -arch arm64`.

Reply:
Let me try to add both `-arch` flags to `qmake`, `qdoc`, `moc` and `uic`

Reply:
Does `x86_64-apple-darwin13.4.0-clang++` support multiple `-arch` flags? I added `-arch x86_64 -arch arm64` to `qmake` flags and yet the output is still `x86_64` only

Reply:
Yes, it supports multiple `-arch` flags.

Reply:
Then I must be doing something wrong at link time, I'll keep looking, but thanks for the information

Reply:
It seems that zlib is causing some errors, at least when trying to compile `rcc`:

```
$BUILD_PREFIX/bin/arm64-apple-darwin20.0.0-clang++ -stdlib=libc++ -arch x86_64 -arch arm64 -headerpad_max_install_names  -arch x86_64 -isysroot /Users/macmini/Documents/
qt-feedstock/sdks/MacOSX11.0.sdk -mmacosx-version-min=10.13 -Wl,-dead_strip -Wl,-rpath,@executable_path/../Frameworks -o ../../../bin/rcc .obj/rcc.o .obj/main.o   $SRC_D
IR/qt-build/qtbase/lib/libQt5Bootstrap.a -framework Foundation -framework CoreServices                                                                                   
Undefined symbols for architecture arm64:                                                                                                                                
  "_z_compress2", referenced from:                                                                                                                                       
      qCompress(unsigned char const*, int, int) in libQt5Bootstrap.a(qbytearray.o)                                                                                       
ld: symbol(s) not found for architecture arm64                                                                                                                           
clang-11: error: linker command failed with exit code 1 (use -v to see invocation)      
```

I tried to use the qt-bundled zlib, but it also fails

Reply:
Do you have a log?

Reply:
I found that we also need to add `-arch x86_64 -arch arm64` to CFLAGS as well

Reply:
I'm unable to compile `qdoc` to support both `arm64` and `x86_64`, since it tries to link against `libc++`. When I set `LLVM_INSTALL_DIR=$PREFIX`, then it complains that there are no symbols for `x86_64`, but if I set `LLVM_INSTALL_DIR=$BUILD_PREFIX`, then the error shows that there are no symbols for  

```
$BUILD_PREFIX/bin/arm64-apple-darwin20.0.0-clang++ -stdlib=libc++ -arch x86_64 -arch arm64 -L$PREFIX/lib -Wl,-rpath,$PREFIX/lib -headerpad_max_install_names  -arch x86_64 -isysroot /Users/macmini/Documents/qt-feedstock/sdks/MacOSX11.0.sdk -mmacosx-version-min=10.13 -Wl,-dead_strip -Wl,-rpath,$PREFIX/lib -Wl,-rpath,@executable_path/../Frameworks -o ../../bin/qdoc .obj/atom.o .obj/clangcodeparser.o .obj/codechunk.o .obj/codemarker.o .obj/codeparser.o .obj/config.o .obj/cppcodemarker.o .obj/cppcodeparser.o .obj/doc.o .obj/docbookgenerator.o .obj/editdistance.o .obj/generator.o .obj/helpprojectwriter.o .obj/htmlgenerator.o .obj/location.o .obj/main.o .obj/node.o .obj/openedlist.o .obj/parameters.o .obj/puredocparser.o .obj/qdocdatabase.o .obj/qdoctagfiles.o .obj/qdocindexfiles.o .obj/quoter.o .obj/sections.o .obj/separator.o .obj/text.o .obj/tokenizer.o .obj/tree.o .obj/xmlgenerator.o .obj/yyindent.o .obj/webxmlgenerator.o .obj/qdoccommandlineparser.o .obj/utilities.o .obj/jscodemarker.o .obj/qmlcodemarker.o .obj/qmlcodeparser.o .obj/qmlmarkupvisitor.o .obj/qmlvisitor.o   -L$PREFIX/lib -lclang $SRC_DIR/qt-build/qtdeclarative/lib/libQt5QmlDevTools.a $SRC_DIR/qt-build/qtbase/lib/libQt5Bootstrap.a -framework Foundation -framework CoreServices
ld: warning: ignoring file /Users/macmini/Documents/qt-feedstock/miniforge3/conda-bld/qt_1637343279362/_h_env_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_pl/lib/libc++.dylib, building for macOS-x86_64 but attempting to link with file built for macOS-arm64
ld: warning: ignoring file /Users/macmini/Documents/qt-feedstock/miniforge3/conda-bld/qt_1637343279362/_h_env_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_pl/lib/libclang.dylib, building for macOS-x86_64 but attempting to link with file built for macOS-arm64
Undefined symbols for architecture x86_64:
  "___gxx_personality_v0", referenced from:
      -[QMacAutoReleasePoolTracker dealloc] in libQt5Bootstrap.a(qcore_mac.o)
      QMacAutoReleasePool::~QMacAutoReleasePool() in libQt5Bootstrap.a(qcore_mac.o)
      QMacVersion::applicationVersion() in libQt5Bootstrap.a(qcore_mac.o)
      QMacVersion::libraryVersion() in libQt5Bootstrap.a(qcore_mac.o)
      Dwarf Exception Unwind Info (__eh_frame) in libQt5Bootstrap.a(qcore_mac.o)
      QString::fromCFString(__CFString const*) in libQt5Bootstrap.a(qcore_foundation.o)
      QString::fromNSString(NSString const*) in libQt5Bootstrap.a(qcore_foundation.o)
```

Reply:
> Is there any way to link against both libc++?

Yes, you can do `lipo -create $BUILD_PREFIX/lib/libc++.dylib $PREFIX/lib/libc++.dylib -output $PREFIX/lib/libc++.dylib` before building qt.